### DISCUSSION TITLE
Adjoint run for vertical coupler

### DISCUSSION THREAD

**Discussion by jsco7509** [NanoComp/meep#2798] [Q&A]
How do I set up the _adjoint run_ for a vertical coupler? This is a 3D simulation attempting to maximise the out coupling efficiency from a waveguide to free space using a vertical coupler. I believe certain reciprocity conditions are not met, and thus, the adjoint run breaks down. The following warning message appears at the start of the adjoint run "output seems independent of input". Thanks in advance.

**Comment by smartalecH** [comment]
Hmm you shouldn't have to set up the adjoint run... meep takes care of that for you. You just need to properly define the optimization problem itself.

In the case of a vertical coupler, you can indeed leverage reciprocity. You can either launch a fiber mode, and maximize power in the waveguide, or launch a waveguide mode and maximize power in the fiber (or far-field or whatever).

The "adjoint" source here will depend on your formulation (as will your "forward" source). But meep is smart enough to place that adjoint source for you if you set up a proper figure of merit and forward run.

For one applied example that used meep to do exactly this, you can see this paper: https://doi.org/10.1364/OE.466015

> The following warning message appears at the start of the adjoint run "output seems independent of input"

This just means your figure of merit does a particular `ObjectiveQuantity` you are passing in.
